qos/evm/check_archival.go

```diff
@@ -0,0 +1,84 @@
+package evm
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/buildwithgrove/path/qos/jsonrpc"
+)
+
+// methodGetBalance is the JSON-RPC method for getting the balance of an account at a specific block number.
+// Reference: https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance
+const methodGetBalance = jsonrpc.Method("eth_getBalance")
+
+// TODO_IN_THIS_PR(@commoddity): determine an appropriate interval for checking archival status.
+const checkArchivalInterval = 20 * time.Minute
+
+var (
+	errNoArchivalBalanceObs      = fmt.Errorf("endpoint has not returned an archival balance response to a %q request", methodGetBalance)
+	errInvalidArchivalBalanceObs = "endpoint has archival balance %s, expected archival balance %s"
+)
+
+// endpointCheckBlockNumber is a check that ensures the endpoint's block height is greater than the perceived block height.
+// It is used to ensure that the endpoint is not behind the chain.
+type endpointCheckArchival struct {
+	// observedArchivalBalance stores the result of processing the endpoint's response
+	// to an `eth_getBalance` request for a specific contract address at a specific block number.
+	observedArchivalBalance string
+	expiresAt               time.Time
+}
+
+// isValid returns an error if the endpoint's block height is less than the perceived block height minus the sync allowance.
+func (e *endpointCheckArchival) isValid(archivalState archivalState) error {
+	if !archivalState.isEnabled() {
+		return nil
+	}
+
+	if e.observedArchivalBalance == "" {
+		return errNoArchivalBalanceObs
+	}
+	if e.observedArchivalBalance != archivalState.expectedBalance {
+		return fmt.Errorf(errInvalidArchivalBalanceObs, e.observedArchivalBalance, archivalState.expectedBalance)
+	}
+
+	return nil
+}
+
+// shouldRun returns true if the check is not yet initialized or has expired.
+func (e *endpointCheckArchival) shouldRun(archivalState archivalState) bool {
```

Comment:
WDYT of moving this to the archivalState itself, making it the sole authority on all archival-related logic, as intended originally?

Reply:
Ditto this comment here: https://github.com/buildwithgrove/path/pull/200#discussion_r2032615193

This method relies on the value `e.expiresAt` which is specific to this endpoint. Unless you're suggesting that I just make it a method on the `archivalState` that takes endpoint as an argument?

Reply:
Moved to service state.